File: src/main/java/adventofcode/jul2020/Day8.java
package adventofcode.jul2020;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Scanner;

public class Day8 {
    public static void main(String[] args) {
        try {
            File file = new File("src\\Input.txt");
            Scanner sc = new Scanner(file);

            HashMap<Integer, String> operations = new HashMap<>();

            int index = 0;


            while (sc.hasNext()) { operations.put(index++, sc.nextLine()); }

            System.out.println("Part one: " + partOne(operations));
            System.out.println("Part two: " + partTwo(operations));

        } catch (FileNotFoundException e) {
            System.out.println("Fuck Java");
        }
    }

    static int partOne(HashMap<Integer, String> operations) {

        HashSet<Integer> visited = new HashSet<>();
        int accumulator = 0;
        int index = 0;

        while (!visited.contains(index)) {

            String op = operations.get(index).split(" ")[0];
            int number = Integer.parseInt(operations.get(index).split(" ")[1]);

            visited.add(index);
            if (op.equals("acc")) { accumulator += number; }
            index += op.equals("jmp") ? number : 1;
        }
        return accumulator;
    }

    static int partTwo(HashMap<Integer, String> operations) {

        HashSet<Integer> visited = new HashSet<>();
        int accumulator = 0;
        int index;

        replace: for (int i = 0; i < operations.size(); i++) {

            Map<Integer, String> operationsCopy = new HashMap<>(operations);

            if (operationsCopy.get(i).startsWith("nop")) {
                operationsCopy.replace(i, operationsCopy.get(i).replace("nop", "acc"));
            } else if (operationsCopy.get(i).startsWith("jmp")) {
                operationsCopy.replace(i, operationsCopy.get(i).replace("jmp", "nop"));
            } else {
                continue;
            }

            index = 0;
            accumulator = 0;
            visited.clear();

            while (!visited.contains(index)) {

                String op = operationsCopy.get(index).split(" ")[0];
                int number = Integer.parseInt(operationsCopy.get(index).split(" ")[1]);
                visited.add(index);
                if (op.equals("acc")) { accumulator += number; }
                index += op.equals("jmp") ? number : 1;

                if (index == operationsCopy.size()) { break replace; }
            }
        }
        return accumulator;
    }
}

File: src/AdventOfCode/Jul2021/Day8/Day8.kt
package AdventOfCode.Jul2021.Day8

import java.io.File

fun main() {

    println("Part one: " +
            File("src/AdventOfCode/Jul2021/Day8/Input")
                .readLines()
                .map { it.split(" | ")[1].split(" ") }
                .sumOf { it.count { c -> c.length == 2 || c.length == 3 || c.length == 4 || c.length == 7 } }
    )

    println("Part two: " + File("src/AdventOfCode/Jul2021/Day8/Input")
        .readLines()
        .map { it.split(" | ") }
        .sumOf { line ->
            val rule = line[0].split(" ")
            val code = line[1].split(" ")
            val map = mutableMapOf<Int, Set<Char>>()
            map[1] = rule.first { it.length == 2 }.toSet()
            map[4] = rule.first { it.length == 4 }.toSet()
            map[7] = rule.first { it.length == 3 }.toSet()
            map[8] = rule.first { it.length == 7 }.toSet()
                    map[3] = rule.first { it.length == 5 && map[1]!!.all { c -> c in it } }.toSet()
                    map[9] = rule.first { it.length == 6 && map[4]!!.all { c -> c in it } }.toSet()
                    map[0] =
                        rule.first { it.length == 6 && map[1]!!.all { c -> c in it } && it.toSet() != map[9] }.toSet()
                    map[6] = rule.first { it.length == 6 && it.toSet() != map[0] && it.toSet() != map[9] }.toSet()
                    map[5] = rule.first { it.length == 5 && it.all { c -> map[6]!!.contains(c) } }.toSet()
                    map[2] = rule.first { it.length == 5 && it.toSet() != map[5] && it.toSet() != map[3] }.toSet()

                    return@sumOf code
                        .map { map.filterValues { value -> it.toSet() == value.toSet() }.keys.first() }
                        .joinToString("")
                        .toInt()
                }
    )
}
